refactor!(readme): introduce jinja

amadaluzia opened the pull request:
Closes #6

## Does this PR break anything?

This PR adds a new means of adding rules and definitions using Jinja and JSON rather than manually recreating the README from scratch. This will invalidate the old way of adding new rules/definitions to the README, ~~though that isn't made explicit yet.~~ It is now explicitly mentioned to add rules/definitions in `sections/`.

## What does this PR add?
This PR adds the ability to read the COC using JSON, as well as a script that allows for templating with Jinja.

## Do I intend to add more to this PR?

~~Yes, do look out until it is not a draft.~~ No.

amadaluzia committed:
refactor!(readme): introduce jinja

amadaluzia committed:
chore(readme): add contributing notes

amadaluzia committed:
chore(pylock): add lock file for better reproducibility

ingenarel commented:
not my call currently, waiting for @maidnaut

djxnzi commented:
This is a really neat and I could see it being useful for a couple things, but it complicates the CoC a bit. It's a lot easier to edit and maintain markdown for the purposes of a CoC. But it is also really useful for being able to parse it through a bot, so I might adopt this for Clang so we can have a !rules command.

We can merge this for now, but I might pull it out into a separate repo for the bot to pull from later.

amadaluzia commented:
> This is a really neat and I could see it being useful for a couple things, but it complicates the CoC a bit. It's a lot easier to edit and maintain markdown for the purposes of a CoC. But it is also really useful for being able to parse it through a bot, so I might adopt this for Clang so we can have a !rules command.

I mean, it is a bit less obvious where you edit the rules of the COC, but I tried making it easy to make new rules with JSON. Furthermore, it is also a bit more tedious to actually apply the changes as well due to needing to have Python on your machine. However, this could be sorted out by using GitHub Actions to regenerate the README.md whenever a new commit is made. I would even think that if GitHub Actions were in place, it would make adding new rules *easier*.

> We can merge this for now, but I might pull it out into a separate repo for the bot to pull from later.

I think this approach be more tedious, as you would have to rewrite the new rule not just in JSON but in the Markdown which you could have generated with one command. This situation is amplified if you use 2 repositories to handle this, as you would have to make two PRs to two repositories rather than one PR to one repository. I personally believe that this approach would be the easiest with some more automation in place but while editing a Markdown file is easier than editing some JSON and regenerating said Markdown file, it's easier than using 2 repositories for each situation.

ingenarel commented:
if the main source is in markdown, we can just simply edit the coc

if it's in json, it's tough for new contributors to know where this rule is that rule is

also discord can render simple markdown, and the coc is in simple markdown (for now at least)

clang can just source the markdown file and simply copy paste it when needed

but if it's in json

then you need to get the json data

then convert to markdown again for clang

i think the json thing just complicates stuff too much, but that's my two cents

amadaluzia commented:
> if the main source is in markdown, we can just simply edit the coc
> if it's in json, it's tough for new contributors to know where this rule is that rule is

It is easier (at the moment) to edit the Markdown directly, I agree, but it is also easy enough to edit JSON. Furthermore, there is a comment embedded in the Jinja template stating that "you are not in the correct place to edit the COC" which I can make more obvious if needed.

> clang can just source the markdown file and simply copy paste it when needed
> but if it's in json
> then you need to get the json data
> then convert to markdown again for clang

Lets run down through the process with Markdown:
- Give clang the rule you want
- Get the Markdown source
- Split the raw source on every new line
- Drop every item until `### 5. Community Rules`
- Take every item until `### Last words:`
- Split the result per `"\n\n"`
- Match clang's input with the first alphanumeric character of any of the rules in the resulting list
- Split the resulting rule from `": "` only once (so you don't end up catching any other `": "`'s)
- Return the 2nd item of the resulting list

Not amazing. Now lets run through the process with JSON:
- Give clang the rule you want
- Get the JSON source
- Parse it into `rules`
- Index `rules[index - 1]` into `rule`
- `return f"**{rule["rule"]}:** {rule["description"]}"`

Markdown is a markup language, not an interchange format. I might just suck at parsing Markdown though.
